Undeclared identifiers `digitalPinToPort` and `digitalPinToBitMask`

When adding Arduino CI to [Adafruit_BusIO](https://github.com/Arduino-CI/Adafruit_BusIO) we get the following output:

```
Located Arduino binary...                              /Applications/Arduino.app
The set of compilers (1) isn't empty...                                        ✓
Checking g++ version... 
    Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
    Apple clang version 12.0.0 (clang-1200.0.32.21)
    Target: x86_64-apple-darwin19.6.0
    Thread model: posix
    InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
...Checking g++ version                                                        ✓
libasan availability for g++...                                             true
Requested unittest platform 'mega2560' is defined in 'platforms' YML...        ✓
Library conforms to Arduino library specification...                         1.5
Unit testing test.cpp with g++... 

Last command:  $ g++ -std=c++0x -o /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/unittest_test.cpp.bin -DARDUINO=100 -D__AVR__ -g -O1 -fno-omit-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -D__AVR_ATmega2560__ -DARDUINO_CI -I/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/vendor/bundle/ruby/2.6.0/bundler/gems/arduino_ci-334b7aa377c8/cpp/arduino -I/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/vendor/bundle/ruby/2.6.0/bundler/gems/arduino_ci-334b7aa377c8/cpp/unittest -I/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/vendor/bundle/ruby/2.6.0/bundler/gems/arduino_ci-334b7aa377c8/cpp/arduino/Arduino.cpp /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/vendor/bundle/ruby/2.6.0/bundler/gems/arduino_ci-334b7aa377c8/cpp/arduino/Godmode.cpp /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/vendor/bundle/ruby/2.6.0/bundler/gems/arduino_ci-334b7aa377c8/cpp/arduino/stdlib.cpp /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/vendor/bundle/ruby/2.6.0/bundler/gems/arduino_ci-334b7aa377c8/cpp/unittest/ArduinoUnitTests.cpp /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_BusIO_Register.cpp /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_I2CDevice.cpp /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp /Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/test/test.cpp

/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:51:48: error: use of undeclared identifier 'digitalPinToPort'
  csPort = (BusIO_PortReg *)portOutputRegister(digitalPinToPort(cspin));
                                               ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:52:15: error: use of undeclared identifier 'digitalPinToBitMask'
  csPinMask = digitalPinToBitMask(cspin);
              ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:54:52: error: use of undeclared identifier 'digitalPinToPort'
    mosiPort = (BusIO_PortReg *)portOutputRegister(digitalPinToPort(mosipin));
                                                   ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:55:19: error: use of undeclared identifier 'digitalPinToBitMask'
    mosiPinMask = digitalPinToBitMask(mosipin);
                  ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:58:51: error: use of undeclared identifier 'digitalPinToPort'
    misoPort = (BusIO_PortReg *)portInputRegister(digitalPinToPort(misopin));
                                                  ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:59:19: error: use of undeclared identifier 'digitalPinToBitMask'
    misoPinMask = digitalPinToBitMask(misopin);
                  ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:61:49: error: use of undeclared identifier 'digitalPinToPort'
  clkPort = (BusIO_PortReg *)portOutputRegister(digitalPinToPort(sckpin));
                                                ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:62:16: error: use of undeclared identifier 'digitalPinToBitMask'
  clkPinMask = digitalPinToBitMask(sckpin);
               ^
8 errors generated.
...Unit testing test.cpp with g++                                              ✗
Skipping compilation of examples...                as requested via command line
Failures: 1
```

Are these things that should be part of the mock Arduino?

The test case is [here](https://github.com/Arduino-CI/Adafruit_BusIO/blob/master/scripts/test.sh).

I tried adding the following to the top of `Adafruit_SPIDevice.cpp`:

```
#define digitalPinToBitMask(P) (P)
#define digitalPinToPort(P) (P)
```

And now we have the following:

```
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:54:29: error: use of undeclared identifier 'portOutputRegister'
  csPort = (BusIO_PortReg *)portOutputRegister(digitalPinToPort(cspin));
                            ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:57:33: error: use of undeclared identifier 'portOutputRegister'
    mosiPort = (BusIO_PortReg *)portOutputRegister(digitalPinToPort(mosipin));
                                ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:61:33: error: use of undeclared identifier 'portInputRegister'
    misoPort = (BusIO_PortReg *)portInputRegister(digitalPinToPort(misopin));
                                ^
/Users/jfoster/Documents/Arduino/libraries/Adafruit_BusIO/src/Adafruit_SPIDevice.cpp:64:30: error: use of undeclared identifier 'portOutputRegister'
  clkPort = (BusIO_PortReg *)portOutputRegister(digitalPinToPort(sckpin));
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 ### Added
 - Add `__AVR__` to defines when compiling
+- Add support for `diditalPinToPort()`, `digitalPinToBitMask()`, and `portOutputRegister()`
 
 ### Changed
 - Move repository from https://github.com/ianfixes/arduino_ci to https://github.com/Arduino-CI/arduino_ci

--- a/SampleProjects/TestSomething/test/outputRegister.cpp
+++ b/SampleProjects/TestSomething/test/outputRegister.cpp
@@ -1,0 +1,23 @@
+#include <ArduinoUnitTests.h>
+#include <Arduino.h>
+
+// https://github.com/arduino-libraries/Ethernet/blob/master/src/utility/w5100.h#L337
+
+unittest(test)
+{
+  uint8_t ss_pin = 12;
+  uint8_t ss_port = digitalPinToPort(ss_pin);
+  assertEqual(12, ss_port);
+  uint8_t *ss_pin_reg = portOutputRegister(ss_port);
+  assertEqual(GODMODE()->pMmapPort(ss_port), ss_pin_reg);
+  uint8_t ss_pin_mask = digitalPinToBitMask(ss_pin);
+  assertEqual(1, ss_pin_mask);
+
+  assertEqual((int) 1, (int) *ss_pin_reg);    // verify initial value
+  *(ss_pin_reg) &= ~ss_pin_mask;              // set SS
+  assertEqual((int) 0, (int) *ss_pin_reg);    // verify value
+  *(ss_pin_reg) |= ss_pin_mask;               // clear SS
+  assertEqual((int) 1, (int) *ss_pin_reg);    // verify value
+}
+
+unittest_main()

--- a/SampleProjects/TestSomething/test/outputRegister.cpp
+++ b/SampleProjects/TestSomething/test/outputRegister.cpp
@@ -1,9 +1,11 @@
 #include <ArduinoUnitTests.h>
 #include <Arduino.h>
 
+// added to fix https://github.com/Arduino-CI/arduino_ci/issues/193
 // https://github.com/arduino-libraries/Ethernet/blob/master/src/utility/w5100.h#L337
 
-unittest(test)
+#if defined(__AVR__)
+unittest(portOutputRegister)
 {
   uint8_t ss_pin = 12;
   uint8_t ss_port = digitalPinToPort(ss_pin);
@@ -19,5 +21,6 @@ unittest(test)
   *(ss_pin_reg) |= ss_pin_mask;               // clear SS
   assertEqual((int) 1, (int) *ss_pin_reg);    // verify value
 }
+#endif
 
 unittest_main()

--- a/cpp/arduino/Godmode.h
+++ b/cpp/arduino/Godmode.h
@@ -30,6 +30,14 @@ unsigned long micros();
   #define NUM_SERIAL_PORTS 0
 #endif
 
+// These definitions allow the following to compile (see issue #193):
+// https://github.com/arduino-libraries/Ethernet/blob/master/src/utility/w5100.h:341
+// add padding because some boards (__MK20DX128__) offset from the given address
+#define MMAP_PORTS_SIZE (MOCK_PINS_COUNT + 256)
+#define digitalPinToBitMask(pin)  (1)
+#define digitalPinToPort(pin)     (pin)
+#define portOutputRegister(port)  (GODMODE()->pMmapPort(port))
+
 class GodmodeState {
   private:
     struct PortDef {
@@ -42,6 +50,8 @@ class GodmodeState {
       bool attached;
       uint8_t mode;
     };
+
+    uint8_t mmapPorts[MMAP_PORTS_SIZE];
 
     static GodmodeState* instance;
 
@@ -87,12 +97,19 @@ class GodmodeState {
       spi.readDelayMicros = 0;
     }
 
+    void resetMmapPorts() {
+      for (int i = 0; i < MMAP_PORTS_SIZE; ++i) {
+        mmapPorts[i] = 1;
+      }
+    }
+
     void reset() {
       resetClock();
       resetPins();
       resetInterrupts();
       resetPorts();
       resetSPI();
+      resetMmapPorts();
       seed = 1;
     }
 
@@ -111,6 +128,9 @@ class GodmodeState {
     static unsigned long getMicros() {
       return instance->micros;
     }
+
+    uint8_t* pMmapPort(uint8_t port) { return &mmapPorts[port]; }
+    uint8_t mmapPortValue(uint8_t port) { return mmapPorts[port]; }
 
     // C++ 11, declare as public for better compiler error messages
     GodmodeState(GodmodeState const&) = delete;

--- a/cpp/arduino/Godmode.h
+++ b/cpp/arduino/Godmode.h
@@ -30,14 +30,6 @@ unsigned long micros();
   #define NUM_SERIAL_PORTS 0
 #endif
 
-// These definitions allow the following to compile (see issue #193):
-// https://github.com/arduino-libraries/Ethernet/blob/master/src/utility/w5100.h:341
-// add padding because some boards (__MK20DX128__) offset from the given address
-#define MMAP_PORTS_SIZE (MOCK_PINS_COUNT + 256)
-#define digitalPinToBitMask(pin)  (1)
-#define digitalPinToPort(pin)     (pin)
-#define portOutputRegister(port)  (GODMODE()->pMmapPort(port))
-
 class GodmodeState {
   private:
     struct PortDef {
@@ -51,7 +43,7 @@ class GodmodeState {
       uint8_t mode;
     };
 
-    uint8_t mmapPorts[MMAP_PORTS_SIZE];
+    uint8_t mmapPorts[MOCK_PINS_COUNT];
 
     static GodmodeState* instance;
 
@@ -98,7 +90,7 @@ class GodmodeState {
     }
 
     void resetMmapPorts() {
-      for (int i = 0; i < MMAP_PORTS_SIZE; ++i) {
+      for (int i = 0; i < MOCK_PINS_COUNT; ++i) {
         mmapPorts[i] = 1;
       }
     }
@@ -158,6 +150,17 @@ void detachInterrupt(uint8_t interrupt);
 // TODO: issue #26 to track the commanded state here
 inline void tone(uint8_t _pin, unsigned int frequency, unsigned long duration = 0) {}
 inline void noTone(uint8_t _pin) {}
+
+// These definitions allow the following to compile (see issue #193):
+// https://github.com/arduino-libraries/Ethernet/blob/master/src/utility/w5100.h:341
+// we allow one byte per port which "wastes" 224 bytes, but makes the code easier
+#if defined(__AVR__)
+  #define digitalPinToBitMask(pin)  (1)
+  #define digitalPinToPort(pin)     (pin)
+  #define portOutputRegister(port)  (GODMODE()->pMmapPort(port))
+#else
+  // we don't (yet) support other boards
+#endif
 
 
 GodmodeState* GODMODE();

